[KOGITO-4858] Automatic build of examples

radtriste opened the pull request:
https://issues.redhat.com/browse/KOGITO-4858

This will parse the features and will generate automatically the needed examples.

Note: This PR may change some examples images names as native is handled separately from profiles.

Many thanks for submitting your Pull Request :heart:! 

Please make sure your PR meets the following requirements:

- [x] You have read the [contributors' guide](https://github.com/kiegroup/kogito-operator/blob/master/README.md#contributing-to-the-kogito-operator)
- [x] Pull Request title is properly formatted: `[KOGITO-XYZ] Subject`
- [x] Pull Request contains a link to the JIRA issue
- [x] Pull Request contains a description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket
- [ ] Your feature/bug fix has a unit test that verifies it
- [ ] You've tested the new feature/bug fix in an actual OpenShift cluster
- [ ] You've added a [RELEASE_NOTES.md](RELEASE_NOTES.md) entry regarding this change

radtriste commented:
First dry_run with command: `make build-examples-images dry_run=true`

```
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"------------------ SHOW SCENARIOS ------------------"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"Feature: Build onboarding-example/hr images"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build onboarding-example/hr image"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build onboarding-example/hr image with native enabled"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"Feature: Build onboarding-example/payroll images"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build onboarding-example/payroll image"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build onboarding-example/payroll image with native enabled"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"Feature: Build process-springboot-example images"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build process-springboot-example image"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build process-springboot-example image with profile(s) persistence"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build process-springboot-example image with profile(s) persistence,events"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"Feature: Build dmn-quarkus-example images"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build dmn-quarkus-example image"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build dmn-quarkus-example image with native enabled"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"Feature: Build dmn-tracing-quarkus images"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build dmn-tracing-quarkus image"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"Feature: Build kogito-travel-agency/extended/travels images"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build kogito-travel-agency/extended/travels image"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build kogito-travel-agency/extended/travels image with native enabled"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"Feature: Build kogito-travel-agency/extended/visas images"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build kogito-travel-agency/extended/visas image"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build kogito-travel-agency/extended/visas image with native enabled"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"Feature: Build process-timer-quarkus images"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build process-timer-quarkus image"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"Feature: Build dmn-drools-quarkus-metrics images"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build dmn-drools-quarkus-metrics image"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"Feature: Build onboarding-example/onboarding-quarkus images"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build onboarding-example/onboarding-quarkus image"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build onboarding-example/onboarding-quarkus image with native enabled"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"Feature: Build onboarding-example/onboarding-springboot images"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build onboarding-example/onboarding-springboot image"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"Feature: Build process-knative-quickstart-quarkus images"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build process-knative-quickstart-quarkus image"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"Feature: Build ruleunit-quarkus-example images"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build ruleunit-quarkus-example image"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"Feature: Build process-mongodb-persistence-quarkus images"}
{"level":"info","T":"2021-04-02T15:08:25.690+0200","logger":"main","msg":"    Scenario: Build process-mongodb-persistence-quarkus image"}
{"level":"info","T":"2021-04-02T15:08:25.691+0200","logger":"main","msg":"    Scenario: Build process-mongodb-persistence-quarkus image with native enabled"}
{"level":"info","T":"2021-04-02T15:08:25.691+0200","logger":"main","msg":"    Scenario: Build process-mongodb-persistence-quarkus image with profile(s) events"}
{"level":"info","T":"2021-04-02T15:08:25.691+0200","logger":"main","msg":"    Scenario: Build process-mongodb-persistence-quarkus image with native enabled and profile(s) events"}
{"level":"info","T":"2021-04-02T15:08:25.691+0200","logger":"main","msg":"Feature: Build process-mongodb-persistence-springboot images"}
{"level":"info","T":"2021-04-02T15:08:25.691+0200","logger":"main","msg":"    Scenario: Build process-mongodb-persistence-springboot image"}
{"level":"info","T":"2021-04-02T15:08:25.691+0200","logger":"main","msg":"    Scenario: Build process-mongodb-persistence-springboot image with profile(s) events"}
{"level":"info","T":"2021-04-02T15:08:25.691+0200","logger":"main","msg":"Feature: Build process-optaplanner-quarkus images"}
{"level":"info","T":"2021-04-02T15:08:25.691+0200","logger":"main","msg":"    Scenario: Build process-optaplanner-quarkus image"}
{"level":"info","T":"2021-04-02T15:08:25.691+0200","logger":"main","msg":"Feature: Build process-timer-springboot images"}
{"level":"info","T":"2021-04-02T15:08:25.691+0200","logger":"main","msg":"    Scenario: Build process-timer-springboot image"}
{"level":"info","T":"2021-04-02T15:08:25.691+0200","logger":"main","msg":"Feature: Build dmn-springboot-example images"}
{"level":"info","T":"2021-04-02T15:08:25.691+0200","logger":"main","msg":"    Scenario: Build dmn-springboot-example image"}
{"level":"info","T":"2021-04-02T15:08:25.691+0200","logger":"main","msg":"Feature: Build process-quarkus-example images"}
{"level":"info","T":"2021-04-02T15:08:25.691+0200","logger":"main","msg":"    Scenario: Build process-quarkus-example image"}
{"level":"info","T":"2021-04-02T15:08:25.691+0200","logger":"main","msg":"    Scenario: Build process-quarkus-example image with native enabled"}
{"level":"info","T":"2021-04-02T15:08:25.691+0200","logger":"main","msg":"    Scenario: Build process-quarkus-example image with profile(s) persistence"}
{"level":"info","T":"2021-04-02T15:08:25.691+0200","logger":"main","msg":"    Scenario: Build process-quarkus-example image with native enabled and profile(s) persistence"}
{"level":"info","T":"2021-04-02T15:08:25.691+0200","logger":"main","msg":"    Scenario: Build process-quarkus-example image with profile(s) persistence,events"}
{"level":"info","T":"2021-04-02T15:08:25.691+0200","logger":"main","msg":"    Scenario: Build process-quarkus-example image with native enabled and profile(s) persistence,events"}
{"level":"info","T":"2021-04-02T15:08:25.691+0200","logger":"main","msg":"------------------ END SHOW SCENARIOS ------------------"}
```

Gonna launch the full testing

radtriste commented on test/steps/mappers/create_maven_options.go:
```diff
@@ -26,14 +26,13 @@ const (
 type MavenCommandConfig struct {
 	Profiles []string
 	Options  []string
+	Native   bool
```
Switch to independent value because `native` is a package type and not a profile (it is for our example but we may change that later)
Also this will make more coherent image names

sutaakar commented:
The code looks good, will wait with final review until the PR is marked as ready.

radtriste commented:
In the latest commit, I cleaned a bit the `test` package because it was mixing features, packages, scripts and main_test.go.

- Moved all previous test go packages to `pkg` folder
- Created `pkg/gherkin`  package for the parsing of features
- Created `pkg/executor` which contains the `bdd_executor.go` file

We now have a cleaner test folder: https://github.com/radtriste/kogito-operator/tree/kogito-4858/test

A lot of files changed due to package renaming.

radtriste commented:
/jenkins test

radtriste commented:
```
$ make build-smoke-examples-images dry_run=true

"level":"info","T":"2021-04-06T21:04:18.096+0200","logger":"main","msg":"Created feature folder = /tmp/examples-features983323429"}
{"level":"info","T":"2021-04-06T21:04:18.105+0200","logger":"main","msg":"------------------ SHOW SCENARIOS ------------------"}
{"level":"info","T":"2021-04-06T21:04:18.105+0200","logger":"main","msg":"Feature: Build dmn-drools-quarkus-metrics images"}
{"level":"info","T":"2021-04-06T21:04:18.105+0200","logger":"main","msg":"    Scenario: Build dmn-drools-quarkus-metrics image"}
{"level":"info","T":"2021-04-06T21:04:18.105+0200","logger":"main","msg":"Feature: Build process-quarkus-example images"}
{"level":"info","T":"2021-04-06T21:04:18.105+0200","logger":"main","msg":"    Scenario: Build process-quarkus-example image"}
{"level":"info","T":"2021-04-06T21:04:18.105+0200","logger":"main","msg":"Feature: Build process-springboot-example images"}
{"level":"info","T":"2021-04-06T21:04:18.105+0200","logger":"main","msg":"    Scenario: Build process-springboot-example image"}
{"level":"info","T":"2021-04-06T21:04:18.105+0200","logger":"main","msg":"------------------ END SHOW SCENARIOS ------------------"}
```

radtriste commented:
/jenkins test

radtriste commented:
/jenkins test

sutaakar reviewed (approved):
Looks good to me, great job

MarianMacik reviewed (approved):
Looks good to me. Good job!

radtriste commented:
/jenkins test

radtriste commented:
/jenkins test
